test/rules/LinearAlgebra/blas.jl

```diff
@@ -0,0 +1,132 @@
+                    Forward, fun, Tret, Const(n), x_annot, Const(inc), y_annot, Const(inc)
+                )
+
+                Tret <: Const && @test ret === ()
+                if Tret <: Duplicated
+                    v, d = ret
+                    @test v ≈ vexp
+                    @test d ≈ dexp
+                elseif Tret <: DuplicatedNoNeed
+                    @test only(ret) ≈ dexp
+                end
+            end
+
+            @testset for Tret in (BatchDuplicated, BatchDuplicatedNoNeed),
+                T in (fun == BLAS.dot ? RTs : RCs)
+
+                batch_size = 3
+                inc = 1
+                x = randn(T, n)
+                y = randn(T, n)
+                ∂xs = ntuple(_ -> randn(T, n), batch_size)
+                ∂ys = ntuple(_ -> randn(T, n), batch_size)
+                vexp = fun(n, x, inc, y, inc)
+                dexp = map(∂xs, ∂ys) do ∂x, ∂y
+                    FiniteDifferences.jvp(
+                        fdm, (x, y) -> fun(n, x, inc, y, inc), (x, ∂x), (y, ∂y)
+                    )[1]
+                end
+                ret = autodiff(
+                    Forward,
+                    fun,
+                    Tret,
+                    Const(n),
+                    BatchDuplicated(x, ∂xs),
+                    Const(inc),
+                    BatchDuplicated(y, ∂ys),
+                    Const(inc),
+                )
+                if Tret <: BatchDuplicated
+                    v, ds = ret
+                    @test v ≈ vexp
+                else
+                    ds = only(ret)
+                end
+                @test all(map(≈, values(ds), dexp))
+            end
+        end
+
+        @testset "reverse" begin
+            fun_overwrite_x = (x, y) -> (s = dot(x, y); fill!(x, 0); s)
+            fun_overwrite_y = (x, y) -> (s = dot(x, y); fill!(y, 0); s)
+
+            @testset for Tret in (Const, Active),
+                Tx in (Const, Duplicated),
+                Ty in (Const, Duplicated),
+                pfun in (identity, pointer),
+                T in (fun == BLAS.dot ? RTs : RCs),
+                (sz, inc) in ((10, 1), ((2, 20), -2)),
+                f in (fun, fun_overwrite_x, fun_overwrite_y)
+
+                Tx <: Const && Ty <: Const && !(Tret <: Const) && continue
+
+                x, ∂x = ntuple(_ -> randn(T, sz), 2)
+                y, ∂y = ntuple(_ -> randn(T, sz), 2)
+                ∂z = randn(T)
+                xcopy, ycopy, ∂xcopy, ∂ycopy = map(copy, (x, y, ∂x, ∂y))
+
+                x_annot =
+                    Tx <: Const ? Const(pfun(x)) : Duplicated(pfun(xcopy), pfun(∂xcopy))
+                y_annot =
+                    Ty <: Const ? Const(pfun(y)) : Duplicated(pfun(ycopy), pfun(∂ycopy))
+
+                vexp = fun(n, x, inc, y, inc)
+                dexp = FiniteDifferences.j′vp(
+                    fdm, (x, y) -> fun(n, x, inc, y, inc), one(T), x, y
+                )
+                ret = autodiff(
```

Comment:
Should use `autodiff_thunk` so that we can test pullback of a non-unit differential, but ran into #791.